enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+
+#### Create/Update
+
+##### Regular DNS name
+
+* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a Namespace and adds rule(s) containing regular
+DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS
+name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.
+* The OVN-K master will then perform DNS lookup for each of the regular DNS names added to the EgressFirewall rules.
+* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookup for each of the
+regular DNS names.
+* As these DNS names have corresponding ``DNSName`` CRs, the ``egressfirewall`` plugin will update the ``.status`` of
+the `DNSName` CRs with the DNS name and the corresponding current IPs along with the TTL and the next lookup time based
+on the TTL. However, this update will only take place if there is a change in the exisiting IP
+addresses or next time to lookup or both for the DNS name.
+* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
+will update the ``AddressSet`` for the DNS name, which is linked with the ACL rule(s) for the corresponding EgressFirewall
+rule(s).
+* The OVN-K master will also receive the response of the DNS lookup query for the DNS name. The OVN-K master will check
+the corresponding ``DNSName`` CR's ``.status`` and if the next lookup time in the status is greater than the next lookup
+time based on the received TTL, then the corresponding CR's ``.status`` will be updated. The corresponding ``AddressSet``
+will also be updated.
+
+##### Wildcard DNS name
+
+* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a Namespace and adds rule(s) containing wildcard
+DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the wildcard DNS names in the EgressFirewall rules, if
+not already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard
+DNS name after replacing the ``*`` with ``wildcard`` (barring any trailinng `.`). For example, if the wildcard DNS name is
+``*.example.com``, then the Name of the corresponding CR will be ``wildcard.example.com``. This is done to adhere to the
+[Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
+The ``.spec.iswildcard`` field of the CR will be set to true, even if it already exists.
+* The OVN-K master will not perform any DNS lookup for the wildcard DNS names directly.
+* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookups from all the pods. If
```

Comment:
Do you mean that this plugin will now intercept all the pods' dns requests?  How does a dns request get directed to the plugin?

Reply:
CoreDNS uses plugin chain to process a DNS query. The plugin chain is defined in the `plugin.cfg` file and which plugins are enabled is defined in the CoreFile. https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/

Reply:
I'm talking about your new plugin.

Reply:
The process of adding it to the plugin chain is describe here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R254-R261